src/ucp/rndv/rndv_ppln.c

```diff
@@ -74,55 +76,79 @@ ucp_proto_rndv_ppln_init(const ucp_proto_init_params_t *init_params)
-    ppln_caps.num_ranges   = 0;
-    ucp_proto_common_add_ppln_range(&ppln_params, frag_range, SIZE_MAX);
 
-    /* Initialize private data */
-    *init_params->priv_size = sizeof(*rpriv);
-    rpriv->frag_proto       = *select_elem;
-    rpriv->frag_size        = frag_max_length;
+    /* Add each proto as a separate variant */
+    ucs_array_for_each(proto, &select_elem->proto_init.protocols) {
+        /* Skip empty variants and reconfig proto */
+        if ((proto->caps.num_ranges == 0) || (proto->proto_id == 0)) {
+            continue;
+        }
+
+        frag_range = proto->caps.ranges + (proto->caps.num_ranges - 1);
+        if (frag_range->max_length == 0) {
+            continue;
+        }
 
-    /* Add ATS overhead */
-    ppln_overhead = ucs_linear_func_make(frag_overhead,
-                                         frag_overhead / frag_max_length);
-    status = ucp_proto_rndv_ack_init(init_params, UCP_PROTO_RNDV_ATS_NAME,
-                                     &ppln_caps, ppln_overhead, &rpriv->ack, 0);
+        ucs_trace("rndv_ppln frag %s" UCP_PROTO_PERF_FUNC_TYPES_FMT,
+                  ucs_memunits_to_str(rpriv->frag_size, frag_size_str,
+                                      sizeof(frag_size_str)),
+                  UCP_PROTO_PERF_FUNC_TYPES_ARG(frag_range->perf));
+
+        frag_proto_priv = &ucs_array_elem(&select_elem->proto_init.priv_buf,
+                                          proto->priv_offset);
+        ucp_proto_rndv_set_variant_config(init_params, proto,
+                                          &sel_param, frag_proto_priv,
+                                          &rpriv->frag_proto_cfg);
+
+        /* Add the single range of the pipeline protocol to ppln_caps */
+        ppln_caps.cfg_thresh = proto->cfg_thresh;
+        ppln_caps.min_length = frag_range->max_length + 1;
+        ppln_caps.num_ranges = 0;
+        ucp_proto_common_add_ppln_range(&ppln_params, frag_range, SIZE_MAX);
+
+        /* Initialize private data */
+        rpriv->frag_size             = frag_range->max_length;
+        rpriv->frag_proto_min_length = proto->caps.min_length;
+
+        /* Add ATS overhead */
+        ppln_overhead = ucs_linear_func_make(
+                frag_overhead, frag_overhead / frag_range->max_length);
+        status        = ucp_proto_rndv_ack_init(
+                init_params, UCP_PROTO_RNDV_ATS_NAME, &ppln_caps,
+                ppln_overhead, &rpriv->ack, 0);
+        if (status != UCS_OK) {
+            break;
+        }
+        ucp_proto_select_caps_cleanup(&ppln_caps);
 
-    ucp_proto_select_caps_cleanup(&ppln_caps);
+        ucp_proto_select_add_proto(init_params, proto->cfg_thresh,
+                                   proto->cfg_priority,
+                                   init_params->caps, init_params->priv,
+                                   sizeof(*rpriv));
+    }
 
-    return status;
+out:
+    if (status != UCS_OK) {
+        ucs_debug("error during rndv ppln probing, status \"%s\"",
```

Comment:
would not it be always unsupported? i'd remove this print

Reply:
It also can be error code from `ucp_proto_rndv_ack_init`